File: dotnet/FileFS.Tests/FileSystemTests.cs
using System.Text;

namespace FileFS.Tests;

public sealed class FileSystemTests
{
    [Fact]
    public void MkfsMountAndGetcwd()
    {
        using var fixture = new TempFsFixture();
        Assert.Equal("/", fixture.FileSystem.Getcwd());
    }

    [Fact]
    public void MkdirAndChdir()
    {
        using var fixture = new TempFsFixture();

        fixture.FileSystem.Mkdir("docs");
        fixture.FileSystem.Chdir("docs");

        Assert.Equal("/docs/", fixture.FileSystem.Getcwd());
    }

    [Fact]
    public void OpenWriteReadRoundtrip()
    {
        using var fixture = new TempFsFixture();

        using (var file = fixture.FileSystem.Open("note.txt", "w"))
        {
            var payload = Encoding.ASCII.GetBytes("hello filefs");
            Assert.Equal(payload.Length, fixture.FileSystem.Write(file, payload));
        }

        using var reader = fixture.FileSystem.Open("note.txt", "r");
        var buffer = new byte[64];
        var read = fixture.FileSystem.Read(reader, buffer);
        Assert.Equal("hello filefs", Encoding.ASCII.GetString(buffer, 0, read));
    }

    [Fact]
    public void CopyRenameAndRemove()
    {
        using var fixture = new TempFsFixture();

        using (var file = fixture.FileSystem.Open("note.txt", "w"))
        {
            Assert.Equal(3, fixture.FileSystem.Write(file, "abc"u8));
        }

        fixture.FileSystem.CopyFile("note.txt", "copy.txt");
        fixture.FileSystem.Rename("copy.txt", "renamed.txt");
        fixture.FileSystem.RemoveFile("renamed.txt");

        Assert.False(fixture.FileSystem.FileExists("renamed.txt"));
        Assert.True(fixture.FileSystem.FileExists("note.txt"));
    }

    [Fact]
    public void OpenDirListsDocs()
    {
        using var fixture = new TempFsFixture();

        fixture.FileSystem.Mkdir("docs");

        using var dir = fixture.FileSystem.OpenDir("/");
        var names = new List<string>();
        DirEntry? entry;
        while ((entry = fixture.FileSystem.ReadDir(dir)) is not null)
        {
            names.Add(entry.Name);
        }

        Assert.Contains("docs", names);
    }

    [Fact]
    public void BeginCommitCreatesFile()
    {
        using var fixture = new TempFsFixture();

        Assert.True(fixture.FileSystem.Begin());
        using (var file = fixture.FileSystem.Open("txn.txt", "w"))
        {
            Assert.Equal(1, fixture.FileSystem.Write(file, "x"u8));
        }

        Assert.True(fixture.FileSystem.Commit());
        Assert.True(fixture.FileSystem.FileExists("txn.txt"));
    }

    private sealed class TempFsFixture : IDisposable
    {
        public TempFsFixture()
        {
            ImagePath = Path.GetTempFileName();
            FileSystem.Mkfs(ImagePath);
            FileSystem = new FileSystem();
            FileSystem.Mount(ImagePath);
        }

        public string ImagePath { get; }

        public FileSystem FileSystem { get; }

        public void Dispose()
        {
            FileSystem.Dispose();

            if (File.Exists(ImagePath))
            {
                File.Delete(ImagePath);
            }

            var journalPath = ImagePath + "-j";
            if (File.Exists(journalPath))
            {
                File.Delete(journalPath);
            }
        }
    }
}

File: swift/Tests/FileFSTests/FileSystemTests.swift
import XCTest
@testable import FileFS

final class FileSystemTests: XCTestCase {
    func testMkfsMountAndCwd() throws {
        let tempDir = try makeTempDir()
        defer { try? FileManager.default.removeItem(at: tempDir) }

        let imagePath = tempDir.appendingPathComponent("test.ffs").path
        try FileSystem.mkfs(at: imagePath)

        let fs = FileSystem()
        try fs.mount(at: imagePath)
        defer { fs.umount() }

        XCTAssertTrue(fs.isMounted)
        XCTAssertEqual(fs.getcwd(), "/")
    }

    func testMkdirChdirWriteReadAndDirListing() throws {
        let tempDir = try makeTempDir()
        defer { try? FileManager.default.removeItem(at: tempDir) }

        let imagePath = tempDir.appendingPathComponent("test.ffs").path
        try FileSystem.mkfs(at: imagePath)

        let fs = FileSystem()
        try fs.mount(at: imagePath)
        defer { fs.umount() }

        try fs.mkdir("docs")
        try fs.chdir("docs")
        XCTAssertEqual(fs.getcwd(), "/docs/")

        let writer = try fs.open("note.txt", mode: "w")
        let payload = Array("hello filefs".utf8)
        XCTAssertEqual(try fs.write(writer, from: payload), payload.count)
        fs.close(writer)

        let reader = try fs.open("note.txt", mode: "r")
        var buffer = [UInt8](repeating: 0, count: 64)
        let count = try fs.read(reader, into: &buffer)
        fs.close(reader)
        XCTAssertEqual(Array(buffer.prefix(count)), payload)

        try fs.chdir("/")
        let dir = try fs.openDir("/")
        var names = Set<String>()
        while let entry = fs.readDir(dir) {
            names.insert(entry.name)
        }
        fs.closeDir(dir)

        XCTAssertTrue(names.contains("docs"))
    }

    func testCopyRenameRemoveAndTransactionCommit() throws {
        let tempDir = try makeTempDir()
        defer { try? FileManager.default.removeItem(at: tempDir) }

        let imagePath = tempDir.appendingPathComponent("test.ffs").path
        try FileSystem.mkfs(at: imagePath)

        let fs = FileSystem()
        try fs.mount(at: imagePath)
        defer { fs.umount() }

        try fs.mkdir("/docs")
        let writer = try fs.open("/docs/note.txt", mode: "w")
        XCTAssertEqual(try fs.write(writer, from: Array("copy me".utf8)), 7)
        fs.close(writer)

        try fs.copyFile(from: "/docs/note.txt", to: "/copy.txt")
        XCTAssertTrue(fs.fileExists("/copy.txt"))

        try fs.rename(from: "/copy.txt", to: "/renamed.txt")
        XCTAssertTrue(fs.fileExists("/renamed.txt"))
        XCTAssertFalse(fs.fileExists("/copy.txt"))

        try fs.removeFile("/renamed.txt")
        XCTAssertFalse(fs.fileExists("/renamed.txt"))

        try fs.begin()
        let txnWriter = try fs.open("/txn.txt", mode: "w")
        XCTAssertEqual(try fs.write(txnWriter, from: [0x78]), 1)
        fs.close(txnWriter)
        try fs.commit()

        XCTAssertTrue(fs.fileExists("/txn.txt"))

        fs.umount()
        try fs.mount(at: imagePath)
        XCTAssertTrue(fs.fileExists("/txn.txt"))
    }

    private func makeTempDir() throws -> URL {
        let dir = FileManager.default.temporaryDirectory.appendingPathComponent(UUID().uuidString, isDirectory: true)
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }
}
